src/windows/osd.c

```diff
@@ -505,3 +505,43 @@ void freeifaddrs(struct ifaddrs *ifa)
 		ifa = next;
 	}
 }
+
+ssize_t ofi_sendmsg_socket(SOCKET fd, const struct msghdr *msg, int flags)
+{
+	ssize_t size;
+	int ret, i;
+	WSABUF *wsa_buf;
+
+	wsa_buf = (WSABUF *)calloc(msg->msg_iovlen, sizeof(WSABUF));
+
+	for (i = 0; i < msg->msg_iovlen; i++) {
+		wsa_buf[i].buf = (char *)msg->msg_iov[i].iov_base;
+		wsa_buf[i].len = msg->msg_iov[i].iov_len;
+	}
+
+	ret = WSASend(fd, wsa_buf, msg->msg_iovlen, &size, flags, NULL, NULL);
+	if (ret)
+	size = (ssize_t)ret;
+
+	return size;
+}
+
+ssize_t ofi_recvmsg_socket(SOCKET fd, const struct msghdr *msg, int flags)
+{
+	ssize_t size;
+	int ret,i;
+	WSABUF *wsa_buf;
+
+	wsa_buf = (WSABUF *)calloc(msg->msg_iovlen, sizeof(WSABUF));
+
+	for (i = 0; i < msg->msg_iovlen; i++) {
+		wsa_buf[i].buf = (char *)msg->msg_iov[i].iov_base;
+		wsa_buf[i].len = msg->msg_iov[i].iov_len;
+	}
+
+	ret = WSARecv(fd, wsa_buf, msg->msg_iovlen, &size, flags, NULL, NULL);
```

Comment:
We may use the analog of BSD's recvmsg - [WSARecvMsg](https://msdn.microsoft.com/en-us/library/windows/desktop/ms741687(v=vs.85).aspx)

Reply:
WSASend/RecvMsg calls cannot be used with stream oriented sockets -- only with datagram and raw sockets.  Silly, yes.

Reply:
Oh, really:

_The socket descriptor in the s parameter must be opened with the socket type set to SOCK_DGRAM or SOCK_RAW_

👎